### PR TITLE
Core: Switch usage to DataFileSet / DeleteFileSet

### PR DESCRIPTION
**Pull request by nastra**
```
Usage of DataFileSet / DeleteFileSet
====================================

Benchmark                                       (numFiles)  Mode  Cnt  Score   Error  Units
ReplaceDeleteFilesBenchmark.replaceDeleteFiles       50000    ss    5  0.892 ± 1.837   s/op
ReplaceDeleteFilesBenchmark.replaceDeleteFiles      100000    ss    5  1.544 ± 0.139   s/op
ReplaceDeleteFilesBenchmark.replaceDeleteFiles      500000    ss    5  2.487 ± 0.542   s/op
ReplaceDeleteFilesBenchmark.replaceDeleteFiles     1000000    ss    5  2.534 ± 1.214   s/op
ReplaceDeleteFilesBenchmark.replaceDeleteFiles     2500000    ss    5  6.490 ± 2.507   s/op


main
=====
Benchmark                                       (numFiles)  Mode  Cnt  Score   Error  Units
ReplaceDeleteFilesBenchmark.replaceDeleteFiles       50000    ss    5  0.897 ± 1.884   s/op
ReplaceDeleteFilesBenchmark.replaceDeleteFiles      100000    ss    5  1.518 ± 0.133   s/op
ReplaceDeleteFilesBenchmark.replaceDeleteFiles      500000    ss    5  2.622 ± 0.862   s/op
ReplaceDeleteFilesBenchmark.replaceDeleteFiles     1000000    ss    5  2.540 ± 0.922   s/op
ReplaceDeleteFilesBenchmark.replaceDeleteFiles     2500000    ss    5  6.608 ± 3.326   s/op
```

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/FastAppend.java [hunk 1]
```diff
@@ -215,7 +213,7 @@ private List<ManifestFile> writeNewManifests() throws IOException {
     }
 
     if (newManifests == null && !newFiles.isEmpty()) {
-      this.newManifests = writeDataManifests(newFiles, spec);
+      this.newManifests = writeDataManifests(Lists.newArrayList(newFiles), spec);
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/FastAppend.java:216]: What about modifying `writeDataManifests` to accept `Collection` and moving the list creation to `divide`?

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java [hunk 2]
```diff
@@ -533,4 +531,51 @@ private Pair<InclusiveMetricsEvaluator, StrictMetricsEvaluator> metricsEvaluator
       return metricsEvaluators.get(partition);
     }
   }
+
+  private class FilesToDeleteHolder {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ManifestFilterManager.java:535]: Is there any way we can do this differently? In theory, we can add another abstract method, similar to how we handle manifest writers.

```
protected abstract Set<F> newFileSet();

protected abstract ManifestWriter<F> newManifestWriter(PartitionSpec spec);

protected abstract ManifestReader<F> newManifestReader(ManifestFile manifest);
```

One caveat is calling this method to initialize an instance field. It is considered a bad practice but implementations will be stateless, so it will work. We could pass `Supplier<Set<F>>` but not sure it is better. In either case, we need to find a way not to have both sets of files here. It will also reduce the number of changes.

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java [hunk 3]
```diff
@@ -372,8 +367,14 @@ private boolean manifestHasDeletedFiles(
 
     for (ManifestEntry<F> entry : reader.liveEntries()) {
       F file = entry.file();
+
+      // add path-based delete to set of files to be deleted
+      if (deletePaths.contains(CharSequenceWrapper.wrap(file.path()))) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ManifestFilterManager.java:372]: Why do we wrap? It is `CharSequenceSet`.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 4]
```diff
@@ -81,8 +83,8 @@ abstract class MergingSnapshotProducer<ThisT> extends SnapshotProducer<ThisT> {
 
   // update data
   private final Map<PartitionSpec, List<DataFile>> newDataFilesBySpec = Maps.newHashMap();
-  private final CharSequenceSet newDataFilePaths = CharSequenceSet.empty();
-  private final CharSequenceSet newDeleteFilePaths = CharSequenceSet.empty();
+  private final DataFileSet newDataFiles = DataFileSet.create();
+  private final DeleteFileSet newDeleteFiles = DeleteFileSet.create();
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:87]: Do we need these extra collections? Can't we use sets in `newDataFilesBySpec` and `newDeleteFilesBySpec`?

**Review by amogh-jahagirdar — COMMENTED** [review]
Discussed with @nastra offline, overall feel like the change is good but I think it's worth running some benchmarks as a sanity check. There really shouldn't be much of a change afterwards but it's more so to make sure there's not some regression we're not expecting.

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java [hunk 5]
```diff
@@ -71,6 +72,7 @@ public String partition() {
   private final PartitionSet deleteFilePartitions;
   private final PartitionSet dropPartitions;
   private final CharSequenceSet deletePaths = CharSequenceSet.empty();
+  private final FilesToDeleteHolder filesToDelete = new FilesToDeleteHolder();
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ManifestFilterManager.java:75]: I'd consider calling it `Set<F> deleteFiles` to follow the existing naming convention: 

```
deleteFilePartitions -> a set of partitions of all files marked for deletion
deletePaths -> a set of paths to remove
deleteExpression -> an expression
```

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java [hunk 6]
```diff
@@ -71,6 +72,7 @@ public String partition() {
...
   private Expression deleteExpression = Expressions.alwaysFalse();
   private long minSequenceNumber = 0;
   private boolean hasPathOnlyDeletes = false;
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ManifestFilterManager.java:78]: Can we get rid of `hasPathOnlyDeletes` now? We only use `deletePath` for path-only deletes after this PR. We can simply check if that set is empty.

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java (hunk 3 above)
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ManifestFilterManager.java:372]: Why do we actually add the file to the set? I worry that it becomes indistinguishable after that. I feel it would be more obvious to keep checking `deletePaths.contains(file.path())` in addition to checking the file set. If this is done purely to validate missing paths, I think it won't be enough.

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java [hunk 7]
```diff
@@ -421,7 +421,7 @@ private ManifestFile filterManifestWithDeletedFiles(
                 entry -> {
                   F file = entry.file();
                   boolean markedForDelete =
-                      deletePaths.contains(file.path())
+                      filesToDelete.markedForDelete(file)
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ManifestFilterManager.java:424]: I would check the path and the set separately here.

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java [hunk 8]
```diff
@@ -230,27 +232,20 @@ SnapshotSummary.Builder buildSummary(Iterable<ManifestFile> manifests) {
...
    */
-  @SuppressWarnings("CollectionUndefinedEquality")
   private void validateRequiredDeletes(ManifestFile... manifests) {
     if (failMissingDeletePaths) {
-      CharSequenceSet deletedFiles = deletedFiles(manifests);
-      ValidationException.check(
-          deletedFiles.containsAll(deletePaths),
-          "Missing required files to delete: %s",
-          COMMA.join(Iterables.filter(deletePaths, path -> !deletedFiles.contains(path))));
+      deletedFiles(manifests).validateRequiredDeletes(filesToDelete);
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ManifestFilterManager.java:237]: What if we have a path-only delete that is missing?

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java (hunk 3 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/ManifestFilterManager.java:372]: oops good point, I didn't pay enough attention here and just re-used what the old impl was doing in https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/ManifestFilterManager.java#L444-L445

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java (hunk 6 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/ManifestFilterManager.java:78]: yep we can surely remove `hasPathOnlyDeletes`. I was trying to keep changes to an absolute minimum in this class and leave such non-essential refactorings out of the diff but I'm fine doing it here

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 4 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:87]: I'm handling this already in https://github.com/apache/iceberg/pull/11254/commits/4d42f180a4597d274f331dd5abe32fd840611f69. I just didn't want to introduce too many changes/refactorings as the PR is already quite large

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java (hunk 5 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/ManifestFilterManager.java:75]: as discussed on Slack last week I renamed this class to `PendingDeleteFile` which implements `DeleteFile` (I just didn't want to make this PR too large).

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java (hunk 8 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/ManifestFilterManager.java]: good catch, I was relying on existing tests to catch this but after your comment I realized that there's no test coverage for this available. I added test coverage and fixed this

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java (hunk 3 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/ManifestFilterManager.java]: my goal was to reduce handling of paths vs files in a bunch of places and only deal with actual files and also reduce required error handling in `validateRequiredDeletes()` but as you mentioned in another comment this misses the case when a path-based delete is performed with a non-existing file and there wasn't any test coverage for this. I've added test coverage and rewrote this

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java (hunk 5 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/ManifestFilterManager.java:75]: makes sense, updated

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java (hunk 2 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/ManifestFilterManager.java]: I'm also not a big fan of having `FilesToDeleteHolder` and it was just an experiment to see how the changes would look, but it is indeed probably better to have an abstract `Set<F> newFileSet()` method. I've refactored the code and it should look cleaner now, thanks for the suggestion

File: core/src/main/java/org/apache/iceberg/FastAppend.java (hunk 1 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/FastAppend.java:216]: no objection to that. I'll do the same for `writeDeleteManifests`since we'll need that too in the next PR (https://github.com/apache/iceberg/pull/11254)

**Comment by nastra** [comment]
> Discussed with @nastra offline, overall feel like the change is good but I think it's worth running some benchmarks as a sanity check. There really shouldn't be much of a change afterwards but it's more so to make sure there's not some regression we're not expecting.

@amogh-jahagirdar I added benchmark results to the PR description

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableTest.java [hunk 9]
```diff
@@ -213,7 +213,7 @@ public void testDropTable() throws IOException {
     table.newAppend().appendFile(file1).appendFile(file2).commit();
 
     // delete file2
-    table.newDelete().deleteFile(file2.path()).commit();
+    table.newDelete().deleteFile(file2).commit();
```
Comment (aokolnychyi) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableTest.java:216]: Is this change required? There should be no behavior change, right?

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 10]
```diff
@@ -407,7 +407,7 @@ public void testDropTable() throws IOException {
     table.newAppend().appendFile(file1).appendFile(file2).commit();
 
     // delete file2
-    table.newDelete().deleteFile(file2.path()).commit();
```
Comment (aokolnychyi) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:410]: Same question here.

File: core/src/main/java/org/apache/iceberg/ManifestFilterManager.java [hunk 11]
```diff
@@ -71,9 +70,9 @@ public String partition() {
   private final PartitionSet deleteFilePartitions;
   private final PartitionSet dropPartitions;
   private final CharSequenceSet deletePaths = CharSequenceSet.empty();
+  private final Set<F> deleteFiles = newFileSet();
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ManifestFilterManager.java:73]: Can we co-locate `deleteFiles` and `deleteFilePartitions` as `deleteFilePartitions` essentially contains a set of partitions derived from `deleteFiles`.

**Review by aokolnychyi — APPROVED** [review]

**Comment by aokolnychyi** [comment]
Actually, the comments are minor and I verified that changes in tests are cosmetic locally. I'll go ahead and merge this one to unblock the subsequent PR.

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableTest.java (hunk 9 above)
Comment (nastra) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableTest.java:216]: it's not a required change, but I wanted to move existing places to the non-path-based delete path

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 10 above)
Comment (nastra) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:410]: it's not a required change, but I wanted to move existing places to the non-path-based delete path
